pkg/infra/lifecyclehooks.go

```diff
@@ -0,0 +1,226 @@
+			podLabel: "",
+		}
+	}
+	return job
+}
+
+var _ = Describe("[Feature:Machine] Lifecycle Hooks should", func() {
+	var client runtimeclient.Client
+	var machineSet *machinev1.MachineSet
+	var workload *batchv1.Job
+	var pod corev1.Pod
+
+	BeforeEach(func() {
+		var err error
+
+		By("Creating the machineset")
+		client, err = framework.LoadClient()
+		// Build machine set parameters
+		expectedReplicas := 1
+		machineSetParams := framework.BuildMachineSetParams(client, expectedReplicas)
+		// Create a label for node and add to machine set parameters
+		machineSetParams.Labels[lifecyclehooksWorkerNodeRoleLabel] = ""
+		// Create machine set
+		machineSet, err = framework.CreateMachineSet(client, machineSetParams)
+		Expect(err).ToNot(HaveOccurred())
+		// Wait for machine to be running
+		framework.WaitForMachineSet(client, machineSet.GetName())
+
+		By("Running a workload on the machine")
+		// Run a pod on this machine
+		workloadMemRequest := resource.MustParse("100000000")
+		workload = newWorkLoad(int32(expectedReplicas), workloadMemRequest, lifecyclehooksWorkerNodeRoleLabel, lifecycleHooksTestLabel)
+		Expect(client.Create(context.Background(), workload)).To(Succeed())
+
+		By("Waiting for job pod to start running on machine.")
+		Eventually(func() (bool, error) {
+			jobPodList, err := framework.GetPods(client, map[string]string{lifecycleHooksTestLabel: ""})
+			if err != nil {
+				return false, err
+			}
+			if len(jobPodList.Items) == expectedReplicas {
+				pod = jobPodList.Items[0]
+				return pod.Status.Phase == corev1.PodRunning, nil
+			}
+			return false, nil
+		}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+	})
+
+	AfterEach(func() {
+		By("Deleting the machineset")
+		cascadeDelete := metav1.DeletePropagationForeground
+		Expect(client.Delete(context.Background(), machineSet, &runtimeclient.DeleteOptions{
+			PropagationPolicy: &cascadeDelete,
+		})).To(Succeed())
+		By("Deleting workload job")
+		Expect(client.Delete(context.Background(), workload, &runtimeclient.DeleteOptions{
+			PropagationPolicy: &cascadeDelete,
+		})).To(Succeed())
+	})
+
+	It("pause lifecycle actions when present", func() {
+		machines, err := framework.GetMachinesFromMachineSet(client, machineSet)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machines).To(HaveLen(1))
+		machine := machines[0]
+		podKey := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+		machineKey := types.NamespacedName{Namespace: machine.Namespace, Name: machine.Name}
+
+		By("Setting lifecycle hooks on the machine")
+		predrainHook := machinev1.LifecycleHook{
+			Name:  "cluster-api-actuator-pkg/pre-drainHook",
+			Owner: "cluster-api-actuator-pkg",
+		}
+		preterminateHook := machinev1.LifecycleHook{
+			Name:  "cluster-api-actuator-pkg/pre-terminateHook",
+			Owner: "cluster-api-actuator-pkg",
+		}
+		Eventually(func() (bool, error) {
+			Expect(client.Get(context.Background(), machineKey, machine)).To(Succeed())
```

Comment:
You can't make assertions directly like this, you either need to return the error from this function, or pass a `g` into the function and make assertions on that.

Currently if this fails once, it will fail the whole test

Check https://onsi.github.io/gomega/#category-3-making-assertions-eminem-the-function-passed-into-codeeventuallycode for an example of using assertions inside, else just `if err != nil { return bool, error}` which can be less confusing

Reply:
Oh, changed it in just one place. Another commit incomming.